correct_periscope_drift.py

```diff
@@ -0,0 +1,431 @@
+    if len(evt_ra) != len(evt_dec):
+        raise ValueError("len(evt_ra) != len(evt_dec), {} != {}".format(
+                len(evt_ra), len(evt_dec)))
+
+    # Transform to Earth Centered Inertial
+    eci = radec2eci(evt_ra, evt_dec)
+    # transform from 3 x N to N x 3
+    eci = eci.transpose()
+
+    att_stack = np.repeat(np.array([ra_nom, dec_nom, roll_nom]),
+                          len(evt_ra)).reshape(3, len(evt_ra))
+    # Transforms
+    Ts = equatorial2transform(att_stack[0], att_stack[1], att_stack[2])
+
+    # The position of the events rotated into the frame of the
+    d_aca = np.sum(Ts.transpose(2, 0, 1) * eci, axis=-1).transpose()
+
+    R2A = 206264.81
+    yag = np.arctan2(d_aca[:, 1], d_aca[:, 0]) * R2A
+    zag = np.arctan2(d_aca[:, 2], d_aca[:, 0]) * R2A
+    return yag, zag
+
+
+def time_bins(times, x, nbins=20):
+    h, bins = np.histogram(times, bins=nbins)
+    bin_centers = (bins[:-1] + bins[1:]) / 2.0
+    inds = np.digitize(times, bins) - 1
+    bin_x = []
+    bin_std = []
+    for idx in range(0, nbins):
+        data = x[inds == idx]
+
+        bin_x.append(np.mean(data))
+        bin_std.append(np.std(data)/np.sqrt(len(data)))
+    return bin_centers, np.array(bin_x), np.array(bin_std)
+
+
+def fit(fit_data, evt_times, data_id, opt):
+
+    init_error = 5
+
+    ui.clean()
+    ui.load_arrays(data_id, evt_times - evt_times[0], fit_data,
+                   np.zeros_like(fit_data) + init_error)
+    v2("Fitting a line to the data to get reduced stat errors")
+    # First just fit a line to get reduced errors on this set
+    ui.polynom1d.line
+    ui.set_model(data_id, 'line')
+    ui.thaw('line.c1')
+    ui.fit(data_id)
+    fit = ui.get_fit_results()
+    calc_error = init_error * np.sqrt(fit.rstat)
+    ui.set_staterror(data_id, calc_error)
+    # Then fit the specified model
+    v2("Fitting a polynomial of degree {} to the data".format(opt['corr_poly_degree']))
+    ui.polynom1d.fitpoly
+    ui.freeze('fitpoly')
+    # Thaw the coefficients requested by the degree of the desired polynomial
+    for deg in range(1, 1 + opt['corr_poly_degree']):
+        ui.thaw("fitpoly.c{}".format(deg))
+    if opt['corr_poly_degree'] > 1:
+        ui.thaw('fitpoly.offset')
+    ui.set_model(data_id, 'fitpoly')
+    ui.fit(data_id)
+    mp = ui.get_model_plot(data_id)
+    return mp
+
+
+# The '@handle_ciao_errors' decorator will catch any error thrown and
+# display it in a format like that of a CIAO tool, then exit with a
+# non-zero status. So, if at any point in your code you need to exit just
+# raise an error and let handle_ciao_errors bother with the display.
+#
+# This also means that routines can be used from ChIPS/Sherpa/other scripts
+# without having to worry about them calling sys.exit.
+
+
+@handle_ciao_errors(TOOLNAME, VERSION)
+def main(opt):
+    events = extract_events(opt['evtfile'],
+                            opt['x'], opt['y'], opt['radius'])
+
+    evt_ra_nom = events.get_key('RA_NOM').value
```

Comment:
The confusion is often rooted in that the _NOM in the event file(*) isn't the 
same as the _NOM in the asol file :(

The _NOM is just the location of the tangent point -- and it completely (well, 
almost completely) arbitrary.  The _PNT is, at least the way we have 
historically described it, the location of the optical axis [though given the 
other offsets discussions this definition is now somewhat muddled.]  For 
non-multi-obi OBSIDs straight out of the archive, the two will be the same.  For 
multi-obi-obsids, they will/may be different.

When users combine/merge datasets, the _NOM's change as they pick a new common 
tanget point for all the observations to be projected to.


(*) Where by event file, I mean all other files except the asol files.





On 06/20/16 15:01, Jean Connelly wrote:
>
> In correct_periscope_drift.py 
> <https://github.com/sot/correct_periscope_drift/pull/1#discussion_r67746898>:
>
> > +
> > +# The '@handle_ciao_errors' decorator will catch any error thrown and
> > +# display it in a format like that of a CIAO tool, then exit with a
> > +# non-zero status. So, if at any point in your code you need to exit just
> > +# raise an error and let handle_ciao_errors bother with the display.
> > +#
> > +# This also means that routines can be used from ChIPS/Sherpa/other scripts
> > +# without having to worry about them calling sys.exit.
> > +
> > +
> > +@handle_ciao_errors(TOOLNAME, VERSION)
> > +def main(opt):
> > +    events = extract_events(opt['evtfile'],
> > +                            opt['x'], opt['y'], opt['radius'])
> > +
> > +    evt_ra_nom = events.get_key('RA_NOM').value
>
> I'm still not clear about when NOM and PNT are different. I've thus-far only 
> tried to define the workings of the tool for single aspect intervals, and 
> don't quite know what we would want for multi-obi datasets or reprojected 
> files, so I should add those limitations/caveats to the docs.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sot/correct_periscope_drift/pull/1/files/cbe3d98891726c3e493d2afe4b325095778bf928#r67746898>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AGWoI01JDOlRLrmlHJBX-2TpFGxEUE5cks5qNuN5gaJpZM4IyK7D>.
>